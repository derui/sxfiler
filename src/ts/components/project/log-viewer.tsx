import { h } from "preact";

export type Props = {
  entries: string[];
  hidden: boolean;
};

/**
 * make item for a log entry
 */
const makeItem = (entry: string) => {
  let level = "info";

  return (
    <div role="listitem" class="log-viewer__log-item" data-testid="logViewer-logItem" key={entry}>
      <span class="log-viewer__item-level" data-testid="logViewer-itemLevel" data-level={level}>
        [{level.toUpperCase()}]
      </span>
      <span class="log-viewer__item-message" data-testid="logViewer-itemMessage">
        {entry}
      </span>
    </div>
  );
};

export const Component: preact.FunctionComponent<Props> = ({ entries, hidden }) => {
  let components: preact.VNode<any>[] = [];
  if (!hidden) {
    components = entries.map(makeItem);
  }

  return (
    <div class="log-viewer__root" data-testid="logViewer-root" aria-hidden={hidden}>
      {components}
    </div>
  );
};
