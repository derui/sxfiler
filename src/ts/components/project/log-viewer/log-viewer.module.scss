@import "../../ui/commom-style/_colors.scss";
@import "../../ui/commom-style/_variables.scss";
@import "../../ui/commom-style/reset.module.css";

.root {
  composes: list-base from "../../ui/list/list.module.scss";

  background-color: $sol_base03;
  color: $sol_base2;

  height: 100%;
  font-size: 1rem;
  overflow-x: auto;
  overflow-y: scroll;

  font-family: monospace;
}

.item {
  display: flex;
  flex-direction: row;
  padding: $nanoSpace 0;
}

.itemLevel {
  flex: 0 0 4rem;
  white-space: pre;
  margin-right: 1rem;

  &[data-level="info"]{
    color: $sol_blue;
  }

  &[data-level="warning"]{
    color: $sol_orange;
  }

  &[data-level="error"]{
    color: $sol_red;
  }
}

.itemMessage {
  flex: 0 0 auto;
}
