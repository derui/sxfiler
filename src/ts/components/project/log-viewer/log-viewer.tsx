import * as React from "react";

import * as ListItem from "../../ui/list-item/list-item";
import * as List from "../../ui/list/list";
import { MessageNotification, Level } from "../../../domains/message-notification";

// eslint-disable-next-line @typescript-eslint/no-var-requires
const styles: ClassNames = require("./log-viewer.module.scss");

export interface ClassNames {
  root?: string;
  item?: string;
  itemLevel?: string;
  itemMessage?: string;
}

export type Props = {
  entries: MessageNotification[];
  hidden: boolean;
};

/**
 * make item for a log entry
 */
const makeItem = (entry: MessageNotification) => {
  let level = "info";
  switch (entry.level) {
    case Level.Info:
      level = "info";
      break;
    case Level.Warning:
      level = "warning";
      break;
    case Level.Error:
      level = "error";
      break;
  }

  return (
    <ListItem.Component key={entry.id} className={styles.item}>
      <span className={styles.itemLevel} data-level={level}>
        [{level.toUpperCase()}]
      </span>
      <span className={styles.itemMessage}>{entry.body}</span>
    </ListItem.Component>
  );
};

export const Component: React.FC<Props> = ({ entries, hidden }) => {
  let components: React.ReactElement[] = [];
  if (!hidden) {
    components = entries.map(makeItem);
  }

  return (
    <List.Component className={styles.root} aria-hidden={hidden}>
      {components}
    </List.Component>
  );
};
