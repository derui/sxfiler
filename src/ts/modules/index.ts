//#IMPORT INDICATOR
import * as Configuration from "./configuration";
import * as Keymap from "./keymap";
import * as Filer from "./filer";
import * as Completer from "./completer";
import * as Decision from "./decision";
// prettier-ignore
import { combineReducers } from "redux";

// export combined types
// prettier-ignore
export type Actions =
  | Configuration.Actions
  | Keymap.Actions
  | Filer.Actions
  | Completer.Actions
  | Decision.Actions
  ;

// export combined reducer
// prettier-ignore
export const reducer = combineReducers({
  configuration: Configuration.reducer,
  keymap: Keymap.reducer,
  filer: Filer.reducer,
  completer: Completer.reducer,
  decision: Decision.reducer,
})

export const emptyState = {
  configuration: Configuration.emptyState,
  keymap: Keymap.emptyState,
  filer: Filer.emptyState,
  completer: Completer.emptyState,
  decision: Decision.emptyState,
};

// export state type
export type State = ReturnType<typeof reducer>;
