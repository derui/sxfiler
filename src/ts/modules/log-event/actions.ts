import { ActionsType } from "../type";
import { ActionTypes } from "./types";
import { LogEvents } from "./reducer";

// implememt action. Use command `hygen module add:action [name of action]` to add template into this place.
//#ACTION INSERTION INDICATOR
export const send = (events: LogEvents[]) => {
  return { type: ActionTypes.SEND, payload: events };
};

// Do not delete this comment below.
// prettier-ignore
export const actions = {
send,
};

// exporting all actions
export type Actions = ActionsType<typeof ActionTypes, typeof actions>;
