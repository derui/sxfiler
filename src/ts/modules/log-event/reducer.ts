import { ActionTypes } from "./types";
import { Actions } from "./actions";

type LogEvent = {
  timestamp: Date;
};

type MoveItemEvent = LogEvent & {
  kind: "MOVE_EVENT";
  source: string;
  destination: string;
};

type CopyItemEvent = LogEvent & {
  kind: "COPY_EVENT";
  source: string;
  destination: string;
};

type DeleteItemEvent = LogEvent & {
  kind: "DELETE_EVENT";
  fullPath: string;
};

type KeymapReloadEvent = LogEvent & {
  kind: "KEYMAP_RELOAD_EVENT";
};

export type LogEvents = MoveItemEvent | CopyItemEvent | DeleteItemEvent | KeymapReloadEvent;

export namespace LogEventCreators {
  export const createMoveItem = (timestamp: Date, source: string, destination: string): MoveItemEvent => ({
    kind: "MOVE_EVENT",
    timestamp,
    source,
    destination,
  });

  export const createCopyItem = (timestamp: Date, source: string, destination: string): CopyItemEvent => ({
    kind: "COPY_EVENT",
    timestamp,
    source,
    destination,
  });

  export const createDeleteItem = (timestamp: Date, fullPath: string): DeleteItemEvent => ({
    kind: "DELETE_EVENT",
    timestamp,
    fullPath,
  });

  export const createKeymapReload = (timestamp: Date): KeymapReloadEvent => ({
    kind: "KEYMAP_RELOAD_EVENT",
    timestamp,
  });
}

// state of type. Please redefine to what you want.
export type State = {
  events: LogEvents[];
};

export const emptyState: State = { events: [] };

const addEvents = (state: State, events: LogEvents[]) => {
  return Object.freeze({ ...state, events: state.events.concat(events) });
};

export const reducer = (state: State = emptyState, action: Actions): State => {
  switch (action.type) {
    case ActionTypes.SEND:
      return addEvents(state, action.payload);
    default:
      return state;
  }
};
