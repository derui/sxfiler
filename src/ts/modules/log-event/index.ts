import { Actions as Actions$ } from "./actions";
import { State as State$ } from "./reducer";
import { LogEvents as LogEvents_ } from "./reducer";

export type Actions = Actions$;
export type State = State$;

export { ActionTypes } from "./types";
export { actions } from "./actions";
export { reducer, emptyState, LogEventCreators } from "./reducer";

export type LogEvents = LogEvents_;
