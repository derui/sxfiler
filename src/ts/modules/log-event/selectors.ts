import { State } from "./reducer";

export const allEvents = (state: State) => Array.from(state.events);
