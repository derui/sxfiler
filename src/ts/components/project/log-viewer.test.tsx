import { h } from "preact";
import { render } from "preact-render-to-string";

import { Component as T } from "./log-viewer";
import { LogEventCreators } from "@/modules/log-event";

describe("Project", () => {
  describe("Log Viewer", () => {
    it("should render empty", () => {
      const tree = render(<T entries={[]} hidden={false} />);

      expect(tree).toMatchSnapshot();
    });

    it("should render items each level", () => {
      const entries = [
        LogEventCreators.createKeymapReload(new Date(0)),
        LogEventCreators.createDeleteItem(new Date(1), "full path"),
      ];
      const tree = render(<T entries={entries} hidden={false} />);

      expect(tree).toMatchSnapshot();
    });

    it("should render empty when hidden it", () => {
      const tree = render(<T entries={[]} hidden={true} />);

      expect(tree).toMatchSnapshot();
    });
  });
});
