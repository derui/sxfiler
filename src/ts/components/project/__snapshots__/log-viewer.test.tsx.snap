// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Log Viewer should render empty 1`] = `"<div class=\\"log-viewer__root\\" data-testid=\\"logViewer-root\\"></div>"`;

exports[`Project Log Viewer should render empty when hidden it 1`] = `"<div class=\\"log-viewer__root\\" data-testid=\\"logViewer-root\\" aria-hidden></div>"`;

exports[`Project Log Viewer should render items each level 1`] = `"<div class=\\"log-viewer__root\\" data-testid=\\"logViewer-root\\"><div role=\\"listitem\\" class=\\"log-viewer__log-item\\" data-testid=\\"logViewer-logItem\\"><span class=\\"log-viewer__item-level\\" data-testid=\\"logViewer-itemLevel\\" data-level=\\"info\\">[INFO]</span><span class=\\"log-viewer__item-message\\" data-testid=\\"logViewer-itemMessage\\">[1970-01-01T00:00:00.000Z] Keymap reloaded.</span></div><div role=\\"listitem\\" class=\\"log-viewer__log-item\\" data-testid=\\"logViewer-logItem\\"><span class=\\"log-viewer__item-level\\" data-testid=\\"logViewer-itemLevel\\" data-level=\\"info\\">[INFO]</span><span class=\\"log-viewer__item-message\\" data-testid=\\"logViewer-itemMessage\\">[1970-01-01T00:00:00.001Z] Deleted: full path</span></div></div>"`;
